src/devices/bus/ieee488/hp9133.cpp

```diff
@@ -0,0 +1,878 @@
+	}
+
+	if (m_dma_addr >= 256)
+		ret |= IO3_R_FASTDON;
+	LOG("%s: %02x\n", __func__, ret);
+	return ret;
+}
+
+void hp9133_device::dma_ram_w(uint8_t data)
+{
+	LOG("%s: %03x = %02x\n", __func__, m_dma_addr, data);
+	if (m_dma_addr < sizeof(m_dma_ram))
+		m_dma_ram[m_dma_addr++] = data;
+}
+
+uint8_t hp9133_device::dma_ram_r(offs_t offset)
+{
+	uint8_t ret;
+
+	if (machine().side_effects_disabled())
+		return 0;
+	if (m_dma_addr >= sizeof(m_dma_ram))
+		return 0;
+	ret = m_dma_ram[m_dma_addr++];
+	LOG("%s: %03x = %02x\n", __func__, m_dma_addr - 1, ret);
+	return ret;
+}
+
+static int sdh_sector_size(uint8_t sdh)
+{
+	const int sizes[4] = { 256, 512, 1024, 128 };
+
+	return sizes[(sdh >> 5) & 3];
+}
+
+static int sdh_devsel(uint8_t sdh)
+{
+	return (sdh >> 3) & 3;
+}
+
+void hp9133_device::hdc_readwrite_sector(bool write)
+{
+	int sector = m_hdc->read(3);
+	uint32_t lba;
+
+	if (sdh_devsel(m_hdc_sdh))
+		return;
+	if (sdh_sector_size(m_hdc_sdh) != 256)
+		return;
+	if (!m_harddisk->exists())
+		return;
+	const auto &info = m_harddisk->get_info();
+	lba = (m_hdc_cylinder * info.heads + m_head) * info.sectors + sector;
+	LOG("%s: %s cyl %4d, head %d, sector %2d lba %8d\n", __func__,
+	    write ? "WRITE" : "READ ", m_hdc_cylinder, m_head, sector, lba);
+	if (write) {
+		m_harddisk->write(lba, m_dma_ram);
+	} else {
+		m_harddisk->read(lba, m_dma_ram);
+		m_dma_addr = 256;
+	}
+}
+
+void hp9133_device::fdc_intrq_w(int state)
+{
+	m_fdc_irq = state;
+	update_intsel();
+}
+
+void hp9133_device::fdc_drq_w(int state)
+{
+	m_fdc_drq = state;
+	update_intsel();
+}
+
+void hp9133_device::floppy_index_cb(floppy_image_device *floppy, int state)
+{
+	m_fdc->index_callback(floppy, state);
+	if (state)
+		m_floppy_index_int = true;
+	update_intsel();
+}
+
+void hp9133_device::hdc_bcs_cb(int state)
+{
+	m_hdc_bcs = state;
+}
+
+void hp9133_device::hdc_bcr_cb(int state)
+{
+	m_dma_addr = 0;
+}
+
+void hp9133_device::hdc_bdrq_cb(int state)
+{
+	bool is_read = m_hdc_cmd >> 4 == 2;
+	m_hdc_bdrq = state;
+	if (state && is_read)
+		hdc_readwrite_sector(false);
+}
+
+
+void hp9133_device::hdc_wg_cb(int state)
+{
+	bool is_write = m_hdc_cmd >> 4 == 3;
+	m_hdc_wg = state;
+	if (state && is_write)
+		hdc_readwrite_sector(true);
+}
+
+void hp9133_device::hdc_intrq_cb(int state)
+{
+	m_hdc_intrq = state;
+	update_intsel();
+}
+
+uint8_t hp9133_device::hdc_reg_r(offs_t offset)
+{
+	return m_hdc->read(offset);
+}
+
+void hp9133_device::hdc_reg_w(offs_t offset, uint8_t data)
+{
+	m_hdc->write(offset, data);
+	switch (offset) {
+	case 2:
+		m_hdc_sector_count = data;
+		break;
+	case 3:
+		m_hdc_sector_number = data;
+		break;
+	case 4:
+		m_hdc_cylinder &= ~0xff;
+		m_hdc_cylinder |= data;
+		break;
+	case 5:
+		m_hdc_cylinder &= ~0xff00;
+		m_hdc_cylinder |= data << 8;
+		break;
+	case 6:
+		m_hdc_sdh = data;
+		break;
+	case 7:
+		m_hdc_cmd = data;
+		break;
+	}
+}
+
+void hp9133_device::cpu_map(address_map &map)
+{
+	map(0x0000, 0x0007).m("i8291a", FUNC(i8291a_device::map));
```

Comment:
You should be able to use `m_gpib` here rather than the literal tag.

Reply:
Indeed.